File: backend/scoring.py
from typing import Dict, Any, Tuple, List
from datetime import datetime

class TriageScore:
    """
    Data container for a completed triage assessment.
    Stores both the raw TEWS score and the final priority decision.
    """
    def __init__(self, base_score: int, final_score: float, priority: str, 
                 override_reason: str = None, clinician_override: bool = False):
        self.base_score = base_score          # Raw TEWS points before any multiplier
        self.final_score = round(final_score, 1)  # Score after discriminator multiplier
        self.priority = priority              # Final color: Red/Orange/Yellow/Green
        self.override_reason = override_reason    # Why it was overridden (if applicable)
        self.clinician_override = clinician_override  # True if a doctor manually changed it
        self.timestamp = datetime.now().isoformat()   # When the triage was done


def calculate_tews_score(vitals: Dict[str, Any]) -> int:
    """
    Step 1: Calculate the Triage Early Warning Score (TEWS).
    
    Each vital sign is scored 0-3 based on how far it deviates from normal.
    0 = normal, 1 = mild concern, 2 = significant concern, 3 = critical
    Maximum possible score = 17 (all vitals at worst)
    """
    score = 0

    # Heart Rate (normal: 51-100 bpm)
    # Bradycardia (<41) or severe tachycardia (>129) = critical (3pts)
    hr = vitals.get("heart_rate")
    if hr is not None:
        if 51 <= hr <= 100: score += 0
        elif 41 <= hr <= 50 or 101 <= hr <= 110: score += 1
        elif 111 <= hr <= 129: score += 2
        else: score += 3

    # Respiratory Rate (normal: 9-14 breaths/min)
    # Bradypnea (<9) or tachypnea (>29) = critical (3pts)
    rr = vitals.get("respiratory_rate")
    if rr is not None:
        if 9 <= rr <= 14: score += 0
        elif 15 <= rr <= 20: score += 1
        elif 21 <= rr <= 29: score += 2
        else: score += 3

    # Systolic Blood Pressure (normal: 101-199 mmHg)
    # Severe hypotension (<71) = critical (3pts)
    sbp = vitals.get("systolic_bp")
    if sbp is not None:
        if 101 <= sbp <= 199: score += 0
        elif 81 <= sbp <= 100: score += 1
        elif sbp >= 200 or 71 <= sbp <= 80: score += 2
        else: score += 3

    # Oxygen Saturation (normal: >=92%)
    # SpO2 <85% = critical hypoxia (3pts)
    # Note: SpO2 is a discriminator — only RED and normal ranges apply
    spo2 = vitals.get("oxygen_saturation")
    if spo2 is not None:
        if spo2 >= 92: score += 0
        elif 85 <= spo2 <= 91: score += 2
        else: score += 3

    # Temperature (normal: 36.0-38.4°C)
    # Hypothermia (<35°C) or hyperthermia (>39°C) = critical
    temp = vitals.get("temperature")
    if temp is not None:
        if 36.0 <= temp <= 38.4: score += 0
        elif (35.0 <= temp < 36.0) or (38.5 <= temp <= 39.0): score += 1
        else: score += 2

    # Mobility — categorical assessment of patient movement
    # Walking independently = normal (0pts)
    # Needs help walking = concern (1pt)
    # Stretcher/immobile = serious (2pts)
    mobility = str(vitals.get("mobility", "")).strip().lower()
    if "walk" in mobility and "help" not in mobility:
        score += 0
    elif "help" in mobility or "assisted" in mobility:
        score += 1
    else:
        score += 2

    return score


def get_clinical_discriminators(red_flags: Dict[str, bool], age: int = None) -> Tuple[str, str]:
    """
    Step 2: Check for clinical discriminators.
    
    Discriminators are high-risk conditions that OVERRIDE the TEWS score entirely.
    A patient can have normal vitals but still be Red/Orange due to a discriminator.
    This is SATS's safety net — ensuring catastrophic conditions are never missed.
    
    Returns: (priority_color, reason) or (None, None) if no discriminator triggered
    """
    is_pediatric = age is not None and age < 12  # Flag for future pediatric-specific logic

    # RED discriminators — immediate life threat, cannot wait
    if (red_flags.get("active_seizure") or       # Active neurological emergency
        red_flags.get("airway_obstruction") or    # ABC: Airway is first priority
        red_flags.get("severe_respiratory_distress") or  # ABC: Breathing compromise
        red_flags.get("uncontrolled_bleeding") or  # ABC: Circulation failure
        red_flags.get("hypoglycaemia")):           # Brain fuel crisis
        return "Red", "Critical Discriminator (Red)"

    # ORANGE discriminators — high risk, treat within 10 minutes
    if (red_flags.get("chest_pain") or           # Possible heart attack
        red_flags.get("shortness_of_breath") or   # Respiratory compromise
        red_flags.get("high_energy_trauma") or    # Internal injury risk (e.g. car crash)
        red_flags.get("severe_burns") or          # Systemic burn response
        red_flags.get("post_ictal")):             # Post-seizure state
        return "Orange", "High-Risk Discriminator (Orange)"

    return None, None  # No discriminator — proceed with TEWS score


def calculate_triage_score(
    vitals: Dict[str, Any], 
    red_flags: Dict[str, bool], 
    age: int = None,
    clinician_override: str = None,
    override_reason: str = None
) -> Dict:
    """
    Step 3: Main orchestrator — combines TEWS + Discriminators + Clinician Override.
    
    Priority order (highest to lowest):
    1. Clinician Override — senior doctor manually sets priority
    2. Discriminator Override — dangerous condition detected
    3. TEWS Score — standard vital signs scoring
    
    Returns a dict with score, priority color, and override details.
    """
    # Calculate raw TEWS score from vitals
    base_score = calculate_tews_score(vitals)
    
    # Check if any discriminator conditions are present
    discriminator_priority, disc_reason = get_clinical_discriminators(red_flags, age)

    # === PRIORITY 1: Clinician Override ===
    # A senior doctor's judgment always takes precedence over the algorithm
    if clinician_override and clinician_override in ["Red", "Orange", "Yellow", "Green"]:
        return {
            "base_score": base_score,
            "final_score": base_score,
            "priority": clinician_override,
            "override_reason": override_reason or "Senior Clinician Override",
            "clinician_override": True,
            "discriminator_override": False
        }

    # === PRIORITY 2: Discriminator Override ===
    # Critical conditions bypass TEWS — score is doubled to reflect urgency
    if discriminator_priority:
        return {
            "base_score": base_score,
            "final_score": base_score * 2.0,  # Multiplier signals severity to dashboard
            "priority": discriminator_priority,
            "override_reason": disc_reason,
            "clinician_override": False,
            "discriminator_override": True
        }

    # === PRIORITY 3: Standard TEWS-based scoring ===
    # No overrides — use raw score to determine color
    # Thresholds based on SATS documentation
    if base_score >= 7:
        priority = "Red"      # Emergency — immediate treatment
    elif base_score >= 4:
        priority = "Orange"   # Very Urgent — treat within 10 mins
    elif base_score >= 2:
        priority = "Yellow"   # Urgent — treat within 60 mins
    else:
        priority = "Green"    # Routine — treat within 4 hours

    return {
        "base_score": base_score,
        "final_score": base_score,
        "priority": priority,
        "override_reason": None,
        "clinician_override": False,
        "discriminator_override": False
    }


def run_tests():
    """
    Test suite — validates scoring engine against known clinical scenarios.
    Run this file directly to verify the engine is working correctly.
    """
    print("🧪 HealthSync Triage Scoring - Test Cases\n")
    
    tests = [
        {
            "name": "1. Normal Adult Patient",
            "vitals": {
                "heart_rate": 78,
                "respiratory_rate": 14,
                "systolic_bp": 125,
                "oxygen_saturation": 97,
                "temperature": 37.0,
                "mobility": "Walking"
            },
            "red_flags": {},
            "age": 34,
            "expected": "Green"
        },
        {
            "name": "2. High Heart Rate + Shortness of Breath",
            "vitals": {
                "heart_rate": 118,
                "respiratory_rate": 26,
                "systolic_bp": 110,
                "oxygen_saturation": 92,
                "temperature": 37.5,
                "mobility": "Walking with help"
            },
            "red_flags": {"shortness_of_breath": True},
            "age": 45,
            "expected": "Orange"
        },
        {
            "name": "3. Critical - Active Seizure",
            "vitals": {
                "heart_rate": 95,
                "respiratory_rate": 18,
                "systolic_bp": 130,
                "oxygen_saturation": 94,
                "temperature": 36.8,
                "mobility": "Stretcher"
            },
            "red_flags": {"active_seizure": True},
            "age": 28,
            "expected": "Red"
        },
        {
            "name": "4. Paediatric Case - High RR + Low SpO2",
            "vitals": {
                "heart_rate": 140,
                "respiratory_rate": 32,
                "systolic_bp": 95,
                "oxygen_saturation": 88,
                "temperature": 38.2,
                "mobility": "Stretcher"
            },
            "red_flags": {},
            "age": 3,
            "expected": "Red"
        },
        {
            "name": "5. Clinician Override Test",
            "vitals": {
                "heart_rate": 82,
                "respiratory_rate": 16,
                "systolic_bp": 118,
                "oxygen_saturation": 96,
                "temperature": 37.1,
                "mobility": "Walking"
            },
            "red_flags": {},
            "age": 52,
            "clinician_override": "Red",
            "override_reason": "Suspected internal bleeding",
            "expected": "Red"
        },
        {
            "name": "6. Moderate Trauma",
            "vitals": {
                "heart_rate": 105,
                "respiratory_rate": 22,
                "systolic_bp": 135,
                "oxygen_saturation": 95,
                "temperature": 37.0,
                "mobility": "Walking with help"
            },
            "red_flags": {"high_energy_trauma": True},
            "age": 31,
            "expected": "Orange"
        }
    ]

    for test in tests:
        result = calculate_triage_score(
            vitals=test["vitals"],
            red_flags=test.get("red_flags", {}),
            age=test.get("age"),
            clinician_override=test.get("clinician_override"),
            override_reason=test.get("override_reason")
        )
        
        status = "✅ PASS" if result["priority"] == test["expected"] else "❌ FAIL"
        print(f"{status} | {test['name']}")
        print(f"   Priority: {result['priority']} | Base: {result['base_score']} | Final: {result['final_score']}")
        if result.get("override_reason"):
            print(f"   Override: {result['override_reason']}")
        print("-" * 60)


if __name__ == "__main__":
    run_tests()

File: src/lib/scoring.js
// src/lib/scoring.js
// SATS-compliant triage scoring
// Based on: Rosedale et al. (2011), Twomey et al. (2012)
//
// TEWS parameters (7 official):
//   HR, RR, SBP, Temperature, AVPU, Mobility, Trauma
//
// SpO2 is NOT a TEWS parameter — it is handled as a discriminator.
// A SpO2 < 92% triggers an Orange discriminator (acute respiratory compromise),
// consistent with the SATS discriminator list.


// ─── STEP 1: TEWS SCORE ──────────────────────────────────────────────────────

function calculateTEWS(vitals = {}) {
  let score = 0;

  // Heart Rate
  // ≤40 = 2, 41–50 = 1, 51–100 = 0, 101–110 = 1, 111–129 = 2, ≥130 = 3
  const hr = vitals.heart_rate;
  if (hr !== undefined && hr !== null) {
    if (hr <= 40) score += 2;
    else if (hr <= 50) score += 1;
    else if (hr <= 100) score += 0;
    else if (hr <= 110) score += 1;
    else if (hr <= 129) score += 2;
    else score += 3;
  }

  // Respiratory Rate
  // <9 = 2, 9–14 = 0, 15–20 = 0, 21–29 = 2, ≥30 = 3
  // Note: 9–20 both score 0; only extremes score points
  const rr = vitals.respiratory_rate;
  if (rr !== undefined && rr !== null) {
    if (rr < 9) score += 2;
    else if (rr <= 20) score += 0;
    else if (rr <= 29) score += 2;
    else score += 3;
  }

  // Systolic BP
  // ≤70 = 3, 71–80 = 2, 81–100 = 1, 101–199 = 0, ≥200 = 2
  const sbp = vitals.systolic_bp;
  if (sbp !== undefined && sbp !== null) {
    if (sbp <= 70) score += 3;
    else if (sbp <= 80) score += 2;
    else if (sbp <= 100) score += 1;
    else if (sbp <= 199) score += 0;
    else score += 2;
  }

  // Temperature
  // Cold/<35 = 2, 35–38.4 = 0, Hot/≥38.5 = 2
  // Note: official table uses binary hot/cold, not gradient
  const temp = vitals.temperature;
  if (temp !== undefined && temp !== null) {
    if (temp < 35) score += 2;
    else if (temp <= 38.4) score += 0;
    else score += 2;
  }

  // AVPU (Level of Consciousness)
  // Alert = 0, Voice = 1, Pain = 2, Unresponsive = 3
  // Confused is treated as equivalent to Voice (scores 1) per SATS table
  const avpu = String(vitals.avpu || "").trim().toUpperCase();
  if (avpu === "A" || avpu === "ALERT") score += 0;
  else if (avpu === "C" || avpu === "CONFUSED") score += 1;
  else if (avpu === "V" || avpu === "VOICE") score += 1;
  else if (avpu === "P" || avpu === "PAIN") score += 2;
  else if (avpu === "U" || avpu === "UNRESPONSIVE") score += 3;
  // If AVPU not recorded, default 0 — but flag as incomplete

  // Mobility
  // Walking = 0, Walking with help = 1, Stretcher/Immobile = 2
  const mobility = String(vitals.mobility || "").toLowerCase();
  if (mobility.includes("stretcher") || mobility.includes("immobile")) score += 2;
  else if (mobility.includes("help") || mobility.includes("assisted")) score += 1;
  else if (mobility.includes("walk")) score += 0;
  // If not recorded, default 0

  // Trauma
  // No = 0, Yes = 1
  if (vitals.trauma === true || vitals.trauma === "yes") score += 1;

  return score;
}


// ─── STEP 2: TEWS SCORE → COLOUR ─────────────────────────────────────────────

function tewsToColour(score) {
  if (score >= 7) return "Red";
  if (score >= 5) return "Orange";
  if (score >= 3) return "Yellow";
  return "Green";
}


// ─── STEP 3: DISCRIMINATOR CHECK ─────────────────────────────────────────────
//
// Discriminators override the TEWS colour upward only — never downward.
// Organised by upgrade target per the official SATS discriminator list.

function checkDiscriminators(redFlags = {}, vitals = {}) {

  // RED discriminators — immediate, cannot wait
  if (
    redFlags.active_seizure ||
    redFlags.airway_obstruction ||
    redFlags.cardiac_arrest ||
    redFlags.facial_inhalation_burn ||
    redFlags.hypoglycaemia         // glucose < 3 mmol/L
  ) {
    return { colour: "Red", reason: "Critical discriminator — immediate treatment required" };
  }

  // ORANGE discriminators — very urgent, treat within 10 minutes
  if (
    redFlags.shortness_of_breath ||
    redFlags.coughing_blood ||
    redFlags.chest_pain ||
    redFlags.uncontrolled_bleeding ||
    redFlags.focal_neurology ||
    redFlags.reduced_consciousness ||
    redFlags.psychosis_aggression ||
    redFlags.threatened_limb ||
    redFlags.high_energy_trauma ||
    redFlags.burn_over_20_percent ||
    redFlags.burn_electrical ||
    redFlags.burn_circumferential ||
    redFlags.burn_chemical ||
    redFlags.burn_face_inhalation ||
    redFlags.poisoning_overdose ||
    redFlags.vomiting_fresh_blood ||
    redFlags.pregnancy_abdominal_trauma ||
    redFlags.diabetic_glucose_11_ketonuria ||
    redFlags.severe_pain ||
    redFlags.seizure_post_ictal ||
    redFlags.dislocation_other_joint ||
    redFlags.fracture_compound
  ) {
    return { colour: "Orange", reason: "High-risk discriminator — treat within 10 minutes" };
  }

  // SpO2 < 92% — not a TEWS parameter, handled here as respiratory discriminator
  const spo2 = vitals.oxygen_saturation;
  if (spo2 !== undefined && spo2 !== null && spo2 < 92) {
    return { colour: "Orange", reason: "SpO₂ < 92% — acute respiratory compromise" };
  }

  // YELLOW discriminators — urgent, treat within 60 minutes
  if (
    redFlags.controlled_bleeding ||
    redFlags.fracture_closed ||
    redFlags.dislocation_finger_toe ||
    redFlags.burn_other ||
    redFlags.persistent_vomiting ||
    redFlags.abdominal_pain ||
    redFlags.pregnancy_trauma ||
    redFlags.pregnancy_pv_bleed ||
    redFlags.diabetic_glucose_17_no_ketonuria ||
    redFlags.moderate_pain
  ) {
    return { colour: "Yellow", reason: "Urgent discriminator — treat within 60 minutes" };
  }

  return { colour: null, reason: null };
}


// ─── COLOUR RANK HELPER ───────────────────────────────────────────────────────
// Used to ensure discriminators never downgrade a TEWS result

const COLOUR_RANK = { "Green": 0, "Yellow": 1, "Orange": 2, "Red": 3 };

function higherColour(a, b) {
  return COLOUR_RANK[a] >= COLOUR_RANK[b] ? a : b;
}


// ─── MAIN EXPORT ─────────────────────────────────────────────────────────────

export function calculate_triage_score(
  vitals = {},
  redFlags = {},
  age = null,
  clinicianOverride = null,
  overrideReason = ""
) {
  const baseScore = calculateTEWS(vitals);
  const tewsColour = tewsToColour(baseScore);
  const { colour: discColour, reason: discReason } = checkDiscriminators(redFlags, vitals);

  // Discriminator can only upgrade, never downgrade
  const computedColour = discColour
    ? higherColour(tewsColour, discColour)
    : tewsColour;

  const isDiscriminatorOverride = discColour !== null && COLOUR_RANK[discColour] > COLOUR_RANK[tewsColour];

  // Clinician override is always final
  if (clinicianOverride && ["Red", "Orange", "Yellow", "Green"].includes(clinicianOverride)) {
    return {
      base_score: baseScore,
      final_score: baseScore,
      priority: clinicianOverride,
      override_reason: overrideReason || "Senior Clinician Override",
      clinician_override: true,
      discriminator_override: false,
      tews_colour: tewsColour,
      avpu_recorded: !!vitals.avpu,
    };
  }

  return {
    base_score: baseScore,
    final_score: baseScore,
    priority: computedColour,
    override_reason: isDiscriminatorOverride ? discReason : null,
    clinician_override: false,
    discriminator_override: isDiscriminatorOverride,
    tews_colour: tewsColour,
    avpu_recorded: !!vitals.avpu,
  };
}